### ISSUE TITLE
[bitnami/postgresql-ha] pgpool child processes exit with status 256, keep on respawning.

### ISSUE THREAD

**Issue by GytisT** [bitnami/charts#4862]
**Which chart**:
postgresql-ha, chart version 6.3.3

**Describe the bug**
Shortly after installing the chart using default values, child processes of pgpool keep on respawning every few / several seconds after exiting with status 256  
```
2020-12-30 17:08:33: pid 1: LOG:  fork a new child process with pid: 16499
2020-12-30 17:08:43: pid 1: LOG:  child process with pid: 10650 exits with status 256
2020-12-30 17:08:43: pid 1: LOG:  fork a new child process with pid: 16529
2020-12-30 17:08:50: pid 1: LOG:  child process with pid: 10177 exits with status 256
2020-12-30 17:08:50: pid 1: LOG:  fork a new child process with pid: 16553
2020-12-30 17:08:58: pid 1: LOG:  child process with pid: 15607 exits with status 256
2020-12-30 17:08:58: pid 1: LOG:  fork a new child process with pid: 16567
2020-12-30 17:09:13: pid 1: LOG:  child process with pid: 15703 exits with status 256
2020-12-30 17:09:13: pid 1: LOG:  fork a new child process with pid: 16619
2020-12-30 17:09:28: pid 1: LOG:  child process with pid: 12807 exits with status 256
2020-12-30 17:09:28: pid 1: LOG:  fork a new child process with pid: 16656
2020-12-30 17:09:33: pid 1: LOG:  child process with pid: 15167 exits with status 256
2020-12-30 17:09:33: pid 1: LOG:  fork a new child process with pid: 16680
2020-12-30 17:09:40: pid 1: LOG:  child process with pid: 15078 exits with status 256
2020-12-30 17:09:40: pid 1: LOG:  fork a new child process with pid: 16703
2020-12-30 17:10:00: pid 1: LOG:  child process with pid: 15845 exits with status 256
2020-12-30 17:10:00: pid 1: LOG:  fork a new child process with pid: 16760
2020-12-30 17:10:28: pid 1: LOG:  child process with pid: 15926 exits with status 256
2020-12-30 17:10:28: pid 1: LOG:  fork a new child process with pid: 16832
2020-12-30 17:10:30: pid 1: LOG:  child process with pid: 15505 exits with status 256
2020-12-30 17:10:30: pid 1: LOG:  fork a new child process with pid: 16849

```
**To Reproduce**

1. Install chart using default values
2. Check pgpool pod logs using kubectl

**Expected behavior**
Child processes not to exit with status 256 forcing pgpool to respawn a new child.

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
version.BuildInfo{Version:"v3.4.2", GitCommit:"23dd3af5e19a02d4f4baa5b2f242645a1a3af629", GitTreeState:"dirty", GoVersion:"go1.15.5"}
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:50:19Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.5", GitCommit:"e338cf2c6d297aa603b50ad3a301f761b4173aa6", GitTreeState:"clean", BuildDate:"2020-12-09T11:10:32Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Additional context**
To my understanding this is not normal expected behaviour. What's interesting is that not all of the child processes are respawning - I have verified this running SHOW POOL PROCESSES command on pgsql (look at the start_time):  
```
SHOW POOL_PROCESSES;
 pool_pid |     start_time      | database | username |     create_time     | pool_counter 
----------+---------------------+----------+----------+---------------------+--------------
 16997    | 2020-12-30 17:11:20 |          |          |                     | 
 16380    | 2020-12-30 17:07:53 |          |          |                     | 
 18658    | 2020-12-30 17:20:48 |          |          |                     | 
 17005    | 2020-12-30 17:11:23 |          |          |                     | 
 17065    | 2020-12-30 17:11:43 |          |          |                     | 
 17693    | 2020-12-30 17:15:18 |          |          |                     | 
 17383    | 2020-12-30 17:13:30 |          |          |                     | 
 18152    | 2020-12-30 17:17:50 |          |          |                     | 
 17679    | 2020-12-30 17:15:10 |          |          |                     | 
 17888    | 2020-12-30 17:16:20 |          |          |                     | 
 18741    | 2020-12-30 17:21:13 |          |          |                     | 
 17889    | 2020-12-30 17:16:20 |          |          |                     | 
 17454    | 2020-12-30 17:13:53 |          |          |                     | 
 16461    | 2020-12-30 17:08:20 |          |          |                     | 
 17264    | 2020-12-30 17:12:50 |          |          |                     | 
 16619    | 2020-12-30 17:09:13 |          |          |                     | 
 18135    | 2020-12-30 17:17:48 |          |          |                     | 
 17365    | 2020-12-30 17:13:28 |          |          |                     | 
 18011    | 2020-12-30 17:17:03 |          |          |                     | 
 18334    | 2020-12-30 17:18:53 |          |          |                     | 
 17739    | 2020-12-30 17:15:30 | postgres | postgres | 2020-12-30 17:21:01 | 1
 18704    | 2020-12-30 17:21:00 |          |          |                     | 
 17594    | 2020-12-30 17:14:40 |          |          |                     | 
 17710    | 2020-12-30 17:15:20 |          |          |                     | 
 18239    | 2020-12-30 17:18:20 |          |          |                     | 
 16832    | 2020-12-30 17:10:28 |          |          |                     | 
 17959    | 2020-12-30 17:16:48 |          |          |                     | 
 16567    | 2020-12-30 17:08:58 |          |          |                     | 
 18511    | 2020-12-30 17:19:58 |          |          |                     | 
 16703    | 2020-12-30 17:09:40 |          |          |                     | 
 17391    | 2020-12-30 17:13:33 |          |          |                     | 
 18128    | 2020-12-30 17:17:43 |          |          |                     | 
(32 rows)
```
`SHOW POOL_HEALTH_CHECK_STATS` does not show a single failed health check, so perhaps that can be ruled out?
Just to be sure I have also tried setting child_life_time to 0, doesn't change a thing. I believe status code would be 512 in that case, anyway.

**Comment by GytisT** [comment]
I have gotten to the bottom of this issue and to save somebody else countless hours of debugging I'll share a story:

Although service (postgresql-ha, not k8s service) seemed to be working with no problems, based on logs from pgpool pod I have noticed child processes constantly exiting with status code 256 and instantly get respawned. However, it wasn't happening to all child processes.
Whilst searching for information on pgpool exit codes, the only information I have managed to land upon was in a mailing list back from 2016 [here](http://www.sraoss.jp/pipermail/pgpool-general/2016-August/004942.html) which states that exit code 256 signifies "pgpool child is killed because pgpool-II goes down or fail over.". I thought to myself that this is weird, because there is not a single log message or anything else that remotely confirms that to be the case. To me, it looked like children were getting killed automatically for whatever reason. Thus, I have decided to check [pgpool's settings documentation](https://www.pgpool.net/docs/latest/en/html/runtime-config-connection-pooling.html), eventually landing upon `child_life_time` setting, which kills idle child processes, it is not set in the chart (nil), but defaults to 300 s (5 minute) by pgpool. To test if that was the issue I have set chart's setting pgpool.childLifeTime to 0. Unfortunately, that did not solve the issue, some child processes still kept on exiting with status code 256.
Keeping in mind the information I have gotten earlier from that mailing list, I have spent countless more hours debugging, trying all sorts of different configurations, thinking that there was something wrong with either pgpool or postgresql or even my cluster until I have executed shell inside of pgpool's container, examined pgpool's generated configuration and realized that although I have set pgpool.childLifeTime to 0 as a chart setting, that did not get propagated and default 5 minute was set in the generated pgpool's configuration file. To see if that was the problem I have noticed that k8s configMap can be used to configure pgpool (pgpool.configurationCM), where I simply set `child_life_time=0`. Surprise, surprise, child processes were no longer exiting randomly. Success!
Regardless of how happy I was to finally solve this issue, it did not yet make 100% sense as exit status code 256 supposedly meant that either pgpool went down or failover was happening, besides the fact that this was occurring with only some of the children, at random times. It all started to make sense after I have checked livenessProbe of pgpool, which essentially runs a query on pgpool itself. By default 32 child processes get spawned, meaning that livenessProbe is going to hit only one of the child processes causing others to get automatically killed by child_life_time setting due to idleness. At that point, almost everything made sense, except for the exit status code and why `pgpool.childLifeTime = 0` chart setting did not get propagated to pgpool's configuration file.
Exit status code - I have simply attributed to that person from the mailing list either being wrong or later pgpool versions changing exit status codes.
`pgpool.childLifeTime = 0` - not getting propagated to final configuration file, well, that's quite a simple bug in bitnami/postgresql-ha chart. Specifically, this part:
```
{{- if .Values.pgpool.childLifeTime }}
    - name: PGPOOL_CHILD_LIFE_TIME
        value: {{ .Values.pgpool.childLifeTime | quote }}
{{- end }}
```
I am pretty sure, that when childLifeTime is set to 0, everything inside the if block gets skipped, which is not right. 0 is a correct value here. Condition inside `if` should be expanded with additional 'and does not equal to 0'. Not too sure if `| quote` part would work with 0, but perhaps pgpool is smart enough to treat 0 and '0' as the same thing.

tl;dr: this is a normal, expected process of pgpool. Only some of the child processes get killed by the `child_life_time` pgpool's setting because k8s livenessProbe set for pgpool's deployment only hits some of them due to load balancing, others idle out.
However, I have discovered a small bug in the chart, which prevents from setting Values.pgpool.childLifeTime to 0, which is considered a valid value by the pgpool. For this I will create a separate, new issue.

**Comment by marcosbc** [comment]
Thanks a lot for the extensive investigation on your part! It seems like this could be an issue with the chart, so I've added a note to keep track of this for further chart improvements. As for the chart fix suggestion it looks fine but let's continue that discussion in the respective issue.

**Comment by rafariossaa** [comment]
Hi,
  I have just merged a change to fix this issue, could you give it a try ?

**Comment by jonathon2nd** [comment]
We are experiencing this now. Even with a brand new installation of `9.1.1` chart

**Comment by rafariossaa** [comment]
Hi @jonathon2nd ,
  Do you mind opening a new issue ?
